icu4c/source/common/wintz.cpp

```diff
@@ -47,77 +48,60 @@ U_NAMESPACE_BEGIN
+          char gmtOffsetTz[11]; // "Etc/GMT+dd" is 11-char long with a terminal null.
+          snprintf(gmtOffsetTz, 11, "Etc/GMT%+d", -utcOffsetMins / 60);
+          return uprv_strdup(gmtOffsetTz);
+        }
+    }
 
-    // convert from wchar_t* (UTF-16 on Windows) to char* (UTF-8).
-    u_strToUTF8(ISOcode, UPRV_LENGTHOF(ISOcode), nullptr,
-        reinterpret_cast<const UChar*>(ISOcodeW), UPRV_LENGTHOF(ISOcodeW), &status);
+    // tzIDStatus == TIME_ZONE_ID_STANDARD || tzIDStatus == TIME_ZONE_ID_DAYLIGHT ||
+    //  dynamicTZI.DaylightBias == 0)
 
-    LocalUResourceBundlePointer bundle(ures_openDirect(nullptr, "windowsZones", &status));
-    ures_getByKey(bundle.getAlias(), "mapTimezones", bundle.getAlias(), &status);
+    CharString winTZ;
+    UErrorCode status = U_ZERO_ERROR;
+    winTZ.appendInvariantChars(UnicodeString(TRUE, dynamicTZI.TimeZoneKeyName, -1), status)
+        .append('\0', status);
 
-    // convert from wchar_t* (UTF-16 on Windows) to char* (UTF-8).
-    u_strToUTF8(dynamicTZKeyName, UPRV_LENGTHOF(dynamicTZKeyName), nullptr,
-        reinterpret_cast<const UChar*>(dynamicTZI.TimeZoneKeyName), -1, &status);
+    // Map Windows Timezone name (non-localized) to ICU timezone ID (Olson timezone id).
+    LocalUResourceBundlePointer winTZBundle(ures_openDirect(nullptr, "windowsZones", &status));
+    ures_getByKey(winTZBundle.getAlias(), "mapTimezones", winTZBundle.getAlias(), &status);
+    ures_getByKey(winTZBundle.getAlias(), winTZ.data(), winTZBundle.getAlias(), &status);
 
-    if (U_FAILURE(status)) {
+    if (U_FAILURE(status))
         return nullptr;
-    }
 
-    if (dynamicTZI.TimeZoneKeyName[0] != 0) {
-        StackUResourceBundle winTZ;
-        ures_getByKey(bundle.getAlias(), dynamicTZKeyName, winTZ.getAlias(), &status);
-
-        if (U_SUCCESS(status)) {
-            const UChar* icuTZ = nullptr;
-            if (errorCode != 0) {
-                icuTZ = ures_getStringByKey(winTZ.getAlias(), ISOcode, &len, &status);
-            }
-            if (errorCode == 0 || icuTZ == nullptr) {
-                /* fallback to default "001" and reset status */
-                status = U_ZERO_ERROR;
-                icuTZ = ures_getStringByKey(winTZ.getAlias(), "001", &len, &status);
-            }
-
-            if (U_SUCCESS(status)) {
-                int index = 0;
-
-                while (!(*icuTZ == '\0' || *icuTZ == ' ')) {
-                    // time zone IDs only contain ASCII invariant characters.
-                    tmpid[index++] = (char)(*icuTZ++);
-                }
-                tmpid[index] = '\0';
-            }
-        }
+    const UChar* icuTZ16 = nullptr;
+    char regionCode[3] = {}; // 2 letter ISO 3166 country code
+    int geoId = GetUserGeoID(GEOCLASS_NATION);
+    int regionCodeLen = GetGeoInfoA(geoId, GEO_ISO2, regionCode, 3, 0);
+    int32_t tzLen;
+    if (regionCodeLen != 0) {
+        icuTZ16 = ures_getStringByKey(winTZBundle.getAlias(), regionCode, &tzLen, &status);
```

Comment:
On purpose, I used 'A' API to avoid an unnecessary code conversion from UTF-16 to UTF-8. My assumption is that geocode is always ASCII-only. In that case, I don't need to use W API.

Reply:
Oh, very good point! :+1: 

Yes, I believe that you are correct. :)
Since the value for the `GEO_ISO2` is always an "ISO 2-letter country/region code" [[1](https://docs.microsoft.com/en-us/windows/win32/api/winnls/ne-winnls-sysgeotype#constants)], and all the codes fro ISO 3166-1 alpha-2 are ASCII letters [[2](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2)], this would let us completely avoid the need to do any complicated conversion. Very nice!

I suppose we actually could even use the W API as well, and then simply down-cast from `wchar_t` to `char` as well. Though that would require an extra buffer, so using the A API avoids this.

Thanks @jungshik!

Reply:
To be clear, I assumed that [A-Za-z] have the same representation (ASCII repesentation) regardless of Windows codepage.

Reply:
I *think* you are correct, and it seems that the documentation also supports this as well.

From the page here: https://docs.microsoft.com/en-us/windows/win32/intl/code-pages
(I added the bold emphasis below)

> For both Windows code pages and OEM code pages, the code values **0x00 through 0x7F correspond to the 7-bit ASCII character set**. Code values 0x00 through 0x19 and 0x7F always represent standardized control characters and **0x20 through 0x7E represent standardized displayable characters**. Characters represented by the remaining codes, 0x80 through 0xff, vary among character sets.

However, just to be extra sure I will start up an internal thread as well to double-check...

Reply:
Adding some more info to this comment thread:
I confirmed and yes, the A version of the `GetGeoInfoA` API here will return [A-Za-z] ASCII, regardless of the code-page, since EBCIDIC code-pages are not allowed as an ANSI code-page.
So this "trick" will work -- but I think it would be good to expand a bit on the comment here that this isn't really good practice and only works as `GEO_ISO2` are all ASCII.
That said, it was pointed out that calling the A API would be *slower* than calling the W API. The A version of the API will end up calling `MultiByteToWideChar` anyways internally to do the conversion, so it ends up being slow.
(It would probably be faster to call the W version of the API, and then simply cast from `wchar_t` to `char` in a loop to convert the string.)
However, since this doesn't get called all that often, and the strings are small, perhaps it doesn't matter too much -- but maybe something to keep in mind for other uses of A APIs.